### PR TITLE
Remove not needed `exec_insert` in mysql2 adapter

### PR DESCRIPTION
**Pull request by kamipo**
Simply it is sufficient to use the method in the super class.

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb [hunk 1]
```diff
@@ -131,15 +131,11 @@ def execute(sql, name = nil)
       def exec_query(sql, name = 'SQL', binds = [], prepare: false)
         result = execute(sql, name)
         @connection.next_result while @connection.more_results?
-        ActiveRecord::Result.new(result.fields, result.to_a)
+        ActiveRecord::Result.new(result.fields, result.to_a) if result
```
Comment (senny) [inline activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb]: Why is that if suddenly necessary?

Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb]: If a query does not have a result, `execute` returns `nil`:

```
irb(main):002:0> conn.execute("create table t(a int)")
   (43.3ms)  create table t(a int)
=> nil
irb(main):003:0> conn.execute("insert into t values (1)")
   (19.3ms)  insert into t values (1)
=> nil
```

`exec_insert` calls `exec_query`:

https://github.com/kamipo/rails/blob/remove_not_needed_exec_insert/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb#L87-L92

```ruby
      # Executes insert +sql+ statement in the context of this connection using
      # +binds+ as the bind substitutes. +name+ is logged along with
      # the executed +sql+ statement.
      def exec_insert(sql, name, binds, pk = nil, sequence_name = nil)
        exec_query(sql, name, binds)
      end
```

This `if result` is needed.

**Comment by meinac** [comment]
If you remove this method from `Mysql2Adapter` then `exec_insert` calls will hit abstract `DatabaseStatements` and as far as I see return types of this 2 methods are different. There might be some applications depending on this method, I suggest you to show deprecation warning before remove this method.

**Comment by kamipo** [comment]
It is the same result (`nil`) as long as we use `exec_insert` for `INSERT` query. What in the case your suggested situation?

**Comment by meinac** [comment]
I see you've covered the nil case in `exec_query` method so it seem fine.

File: activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb (hunk 1 above)
Comment (senny) [inline activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb]: @kamipo are there tests that cover this case? Are there no callers that rely on the fact that `exec_query` did not return `nil` previously.

Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb]: Previously `exec_query` raises `NoMethodError` when executing no result queries (`INSERT`, `UPDATE`, `DELETE`, and all DDL). I think raising `NoMethodError` is incorrect behavior.

```
irb(main):002:0> conn.execute("create table t(a int)")
   (43.3ms)  create table t(a int)
=> nil
irb(main):003:0> conn.execute("insert into t values (1)")
   (19.3ms)  insert into t values (1)
=> nil
irb(main):004:0> conn.exec_query("insert into t values (1)")
  SQL (28.6ms)  insert into t values (1)
NoMethodError: undefined method `fields' for nil:NilClass
```

File: activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb [hunk 2]
```diff
@@ -131,11 +131,7 @@ def execute(sql, name = nil)
       def exec_query(sql, name = 'SQL', binds = [], prepare: false)
         result = execute(sql, name)
         @connection.next_result while @connection.more_results?
-        ActiveRecord::Result.new(result.fields, result.to_a)
-      end
-
-      def exec_insert(sql, name, binds, pk = nil, sequence_name = nil)
-        execute to_sql(sql, binds), name
+        ActiveRecord::Result.new(result.fields, result.to_a) if result
```
Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb]: Previously `exec_query` raises `NoMethodError` when executing no result queries (`INSERT`, `UPDATE`, `DELETE`, and all DDL). I think raising `NoMethodError` is incorrect behavior.

```
irb(main):002:0> conn.execute("create table t(a int)")
   (43.3ms)  create table t(a int)
=> nil
irb(main):003:0> conn.execute("insert into t values (1)")
   (19.3ms)  insert into t values (1)
=> nil
irb(main):004:0> conn.exec_query("insert into t values (1)")
  SQL (28.6ms)  insert into t values (1)
NoMethodError: undefined method `fields' for nil:NilClass
```

**Comment by kamipo** [comment]
This PR simply remove needless `exec_insert` method now by fixed `NoMethodError` at #23954.

**Comment by senny** [comment]
@kamipo thanks.

**Pull request merged**